src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -875,35 +943,164 @@ object FunctionRegistry {
 
   val aggST = Box[SymbolTable]()
 
+  // FIXME: I suspect I need to pop after the calls to k because it really ought
+  // to be a void method
   registerLambdaAggregatorTransformer("flatMap", { (a: CPS[Any], f: (Any) => Any) =>
-    { (k: Any => Any) => a { x =>
+    { (k: Any => Unit) => a { x =>
       val r = f(x).asInstanceOf[IndexedSeq[Any]]
       var i = 0
       while (i < r.size) {
         k(r(i))
         i += 1
       }
     } }
+  }, { (x: Code[AnyRef], f: Code[AnyRef] => CM[Code[AnyRef]]) =>
+    { (k: Code[AnyRef] => CM[Code[Unit]]) => for (
+      is <- f(x);
+      (str, _r) <- CM.memoize(is);
+      r = Code.checkcast[scala.collection.SeqLike[AnyRef, IndexedSeq[AnyRef]]](_r);
+      (stn, n) <- CM.memoize(Invokeable.lookupMethod[scala.collection.SeqLike[AnyRef, IndexedSeq[AnyRef]], Int]("size", Array()).invoke(r, Array()));
+      i <- CM.newLocal[Int];
+      ri = Invokeable.lookupMethod[scala.collection.SeqLike[AnyRef, IndexedSeq[AnyRef]], AnyRef]("apply", Array(implicitly[ClassTag[Int]].runtimeClass)).invoke(r, Array(i));
```

Comment:
This is really painful, but I don't have an answer other than allowing `invoke` on things that aren't sub types of `AnyRef`.